Extended IRGen support with tensor-typed attrs of general shapes, bool/int/float/string attrs, int list attrs, and input list

mhong opened the pull request:
Passed tensor.swift in eager mode, except for one test under GPU. We will debug
that later.

The main changes (to focus on when reviewing) are in IRGenSIL.cpp. 

Some detailed change notes:

1. Generalized the runtimes functions for creating int and float const tensors to handle arbitrary shapes.

2. Avoided looking up SIL functions via function signatures generated by clang importer, and as such
reverted to change to SILDeserializer::readSILFunctionChecked() needed by that earlier.

3. In eager mode, the "T" attr needs to be explicitly specified, so we added that to a few places in TensorFlow stdlib.

4. Moved some consts and helper functions from TFDeviceSupport/TFUtilities (in SILOptimizer/) to GraphOperationInfo.h (some other place under SIL/ would do too), so that IRGen can use them.

Commit:
Extended IRGen support with tensor-typed attrs of general shapes,
int/float/string attrs, int list attrs, and input list.

Passed tensor.swift in eager mode, except for one test under GPU. We will debug
that later.

Some detailed change notes:

1. Generalized the runtimes functions for creating int and float const tensors to handle arbitrary shapes.

2. Avoided looking up SIL functions via function signatures generated by clang importer, and as such
reverted to change to SILDeserializer::readSILFunctionChecked() needed by that earlier.

3. In eager mode, the "T" attr needs to be explicitly specified, so we added that to a few places in TensorFlow stdlib.

mhong commented:
This PR cannot yet be tested by CI -- we need to pull in a new tensorflow patch that fixes a bug in eager execution.

Commit:
Removed a comment.

Commit:
Fixed a comment.

rxwei commented on lib/IRGen/IRGenSIL.cpp:
```diff
@@ -931,22 +931,24 @@ class IRGenSILFunction :
 
-  // SWIFT_ENABLE_TENSORFLOW
-  // Returns the LLVM function with `funcName`. It must exist.
-  llvm::Function *findFunction(StringRef funcName, SILModule &silModule) {
```
Removing this is great! But I think it deleted the SWIFT_ENABLE_TENSORFLOW that should've stayed here because the code below is TF.

mhong replied:
You are right. brought back this comment.

rxwei commented on lib/IRGen/IRGenSIL.cpp:
```diff
@@ -931,22 +931,24 @@ class IRGenSILFunction :
+  void createArrayAndSize(ArrayRef<SrcT> elements, bool isFloat,
+                          const char *arrayName,
+                          std::function<llvm::Value *(SrcT elt)> convertElt,
+                          Address *array, llvm::Value **numElts);
```
Is it better to use `*&` here? It's a little safer and has a precedent in this file https://github.com/apple/swift/blob/4a9bd582f1d303c6b733ccbf5a3dc5ec44c42a5e/lib/IRGen/IRGenSIL.cpp#L3549, whereas `**` has no precedent.

mhong replied:
The google c++ style is to pass an output param by pointer, so that it is more distinguished from input params. For consistency with local style, I will change to passing by reference, but name the var to be `<foo>Out`.

rxwei replied:
Ok! Good to know.

rxwei commented on lib/IRGen/IRGenSIL.cpp:
```diff
@@ -1916,6 +1918,8 @@ static void abortOnGraphOp(IRGenFunction &IGF, llvm::StringRef errMessage) {
 
 // Create and return a (i8*-typed) address value to a constant string.
+// TODO: As a code size optimization, if we have created this global string
+// before, we can avoid creating a new one.
```
I think that would require defining this as a method in `IRGenSILFunction::`. Would it make sense to do it now? IIUC we may be generate new strings super often, since each op has a device name attribute.

rxwei commented on lib/IRGen/IRGenSIL.cpp:
```diff
@@ -1932,6 +1936,25 @@ static llvm::Value *createStringValAddr(IRGenModule &IGM, StringRef strVal) {
+    std::function<llvm::Value *(SrcT elt)> convertElt, Address *array,
+    llvm::Value **numElts) {
+  *numElts = llvm::ConstantInt::get(IGM.Int32Ty, elements.size());
+  *array = isFloat ? createAlloca(IGM.FloatTy, *numElts,
```
The two `createAlloca`s are identical except the type here. Consider a simpler approach:
```cpp
auto *type = isFloat ? IGM.FloatTy : (llvm::Type *)IGM.Int64Ty;
*array = createAlloca(type, *numElts, IGM.getPointerAlignment(), "tensorEltVals");
```

mhong replied:
Good idea! Done.

rxwei commented on lib/IRGen/IRGenSIL.cpp:
```diff
@@ -1932,6 +1936,25 @@ static llvm::Value *createStringValAddr(IRGenModule &IGM, StringRef strVal) {
+                                  IGM.getPointerAlignment(), "tensorEltVals")
+                   : createAlloca(IGM.Int64Ty, *numElts,
+                                  IGM.getPointerAlignment(), "tensorEltVals");
+  for (unsigned idx = 0; idx < elements.size(); ++idx) {
```
Use an intermediate variable to avoid recomputing `.size()`. 
Or simply `for (auto idx : range(elements.size()))`.

mhong replied:
done

rxwei commented on lib/IRGen/IRGenSIL.cpp:
```diff
@@ -1991,43 +2014,63 @@ void IRGenSILFunction::visitGraphOperationInst(GraphOperationInst *i) {
+    LLVM_DEBUG(llvm::dbgs()
+               << " Adding input of type " << tensorHandle->getType() << ".\n");
+    auto valueKind = classifyTensorFlowValue(tensorHandle->getType());
+    assert(valueKind != TFValueKind::Nope &&
```
You can call `tf::isTensorFlowValue(tensorHandle->getType())`. That function does the same thing as these two lines.

mhong replied:
Nice. done.

rxwei commented on lib/IRGen/IRGenSIL.cpp:
```diff
@@ -1991,43 +2014,63 @@ void IRGenSILFunction::visitGraphOperationInst(GraphOperationInst *i) {
       Builder.CreateCall(TFENewOpFn, {eagerContext, opNameValAddr, status});
   checkOk(status);
 
+  auto addTensorhandleAsOpInput = [=](SILValue tensorHandle) {
```
Hmm, why `[=]`? `IGM` and `Builder` shouldn't be copied. I'm surprised this didn't result into strange behavior or assertion.

mhong replied:
Changed to & to be safer. Perhaps copying them is fine -- If these objects cannot be copied, usually their copy c'tor should be deleted/disabled. But avoiding copying is better.

rxwei replied:
Their copy c'tors should've been deleted. These objects are very, very large structs.

rxwei commented on lib/IRGen/IRGenSIL.cpp:
```diff
@@ -2041,15 +2084,15 @@ void IRGenSILFunction::visitGraphOperationInst(GraphOperationInst *i) {
     switch (attrInfo.second) {
     case GraphOperationInfo::ArgumentLowering::Input:
-      llvm_unreachable("Input classes cannot exist for attributes");
+      assert(0 && "Input classes cannot exist for attributes");
```
What's the reason not to use unreachable?

mhong replied:
Assertion gives us a nice crash stack trace pointing to this source location. llvm_unreachable() does not. Once we complete the impl, using llvm_unreachable should be fine on invariant checks. But in the mean time, using assertion makes it easier to debug test cases.

rxwei commented on lib/IRGen/IRGenSIL.cpp:
```diff
@@ -2059,33 +2102,126 @@ void IRGenSILFunction::visitGraphOperationInst(GraphOperationInst *i) {
+          Address vals;
+          createArrayAndSize<SymbolicValue>(
+              elements, /*isFloat*/ false, "intListAttr",
+              [=](SymbolicValue elt) {
```
`IGM` is being copied here.

mhong replied:
changed to &

rxwei commented on lib/IRGen/IRGenSIL.cpp:
```diff
@@ -2059,33 +2102,126 @@ void IRGenSILFunction::visitGraphOperationInst(GraphOperationInst *i) {
+        } else {
           // In this case, just let eager pick a default device.
-          if (deviceStr == "ALL_DEVICES")
+          if (attrValue == ALL_DEVICES)
```
The variable names for these device name constants ("ALL_DEVICES", "CPU", "GPU", and "TPU") are going to be confusing here because it's some TF-specific code in general IRGen. Consider adding a "TF_" prefix to these variable names.

mhong replied:
That sounds reasonable. I left a TODO, and will tackle it in another PR, so that this PR does not get largeer (and harder to review).

rxwei commented on lib/IRGen/IRGenSIL.cpp:
```diff
@@ -2059,33 +2102,126 @@ void IRGenSILFunction::visitGraphOperationInst(GraphOperationInst *i) {
         auto *setAttrTypeFn = IGM.getTFE_OpSetAttrTypeFn();
         auto dtypeValAddr = createStringValAddr(IGM, attrName.c_str());
         dtypeAttr = convertSwiftTypeToTF(attr.value.getMetatypeValue());
+        LLVM_DEBUG(llvm::dbgs() << " Setting type-typed attr " << attrName
```
Remove the leading space in the message?

mhong replied:
I intentionally added such indentation, to make the debugging msg a bit easier to read.

rxwei commented on lib/IRGen/IRGenSIL.cpp:
```diff
@@ -2096,39 +2232,98 @@ void IRGenSILFunction::visitGraphOperationInst(GraphOperationInst *i) {
+      // Tensor can support two cases: an array case (possibly empty), and a
+      // scalar case.
+      SmallVector<SymbolicValue, 4> elements;
+      bool isFloat;
```
Initialize?

rxwei commented on lib/IRGen/IRGenSIL.cpp:
```diff
@@ -2096,39 +2232,98 @@ void IRGenSILFunction::visitGraphOperationInst(GraphOperationInst *i) {
+                   << "The elt dtype of tensor-typed attr is " << eltType
+                   << ", with tfDtype = " << tfDtype << ".\n");
+        // 1 means TF_FLOAT.
+        isFloat = (tfDtype == 1);
```
Redundant parentheses.

rxwei commented on lib/IRGen/IRGenSIL.cpp:
```diff
@@ -2096,39 +2232,98 @@ void IRGenSILFunction::visitGraphOperationInst(GraphOperationInst *i) {
+      Address tensorEltVals;
+      createArrayAndSize<SymbolicValue>(
+          elements, isFloat, "tensorEltVals",
+          [=](SymbolicValue elt) {
```
This is copying `IGM`.

rxwei commented on lib/Serialization/DeserializeSIL.cpp:
```diff
@@ -430,9 +430,8 @@ SILDeserializer::readSILFunctionChecked(DeclID FID, SILFunction *existingFn,
     break;
     
   case SILStage::Lowered:
-    if (!declarationOnly) // SWIFT_ENABLE_TENSORFLOW
```
Awesome.

rxwei commented on stdlib/public/CTensorFlow/ctensorflow_init.cpp:
```diff
@@ -34,52 +34,97 @@ void InitTensorFlowRuntime(unsigned char enable_debug_logging,
+void *swift_tfc_CreateFloatTensor(int32_t num_dims, int64_t *dims, float *vals,
+                                  TF_Status *status) {
+  // Compute the total memory size of the tensor value.
+  unsigned long long totalElements = 1;
```
[TensorFlow's tensor byte size](https://github.com/tensorflow/tensorflow/blob/3b94d75a9e10ef8ef33760d0ef6aad326e1353ba/tensorflow/c/c_api.h#L270) representation is `size_t`, not ULL. They are often different I think.

rxwei commented on stdlib/public/CTensorFlow/ctensorflow_init.cpp:
```diff
@@ -34,52 +34,97 @@ void InitTensorFlowRuntime(unsigned char enable_debug_logging,
+
+  // Compute the total memory size of the tensor value.
+  // totalElements can be 0 if shape is [] (i.e., num_dims = 1, dims[0] = 0).
+  unsigned long long totalElements = 1;
```
Same here.

rxwei commented on stdlib/public/TensorFlow/Tensor.swift:
```diff
@@ -447,7 +449,10 @@ extension TensorElementLiteral : ExpressibleByArrayLiteral {
   @inlinable @inline(__always)
   public init(arrayLiteral elements: TensorElementLiteral<Scalar>...) {
-    let handle: TensorHandle<Scalar> = #tfop("Pack", elements)
+    // Attr T (non-optional in the op definition) need not be specified when we
```
Is this a workaround for your TF patch? Will we be able to eliminate this chance after you pull from TF?

rxwei commented:
> Avoided looking up SIL functions via function signatures generated by clang importer, and as such
reverted to change to SILDeserializer::readSILFunctionChecked() needed by that earlier.

This is great.

> In eager mode, the "T" attr needs to be explicitly specified, so we added that to a few places in TensorFlow stdlib.

Can this change be avoided after your TF patch is applied?

Commit:
Updated TF repo hash to bring in the eager execution bug fix on default attrs: https://github.com/tensorflow/tensorflow/commit/d493a7f2fdbbc29a292741135f4c1598352e876b.

With this change, I have to turn off "export TF_NCCL_VERSION=1" in my local env
to get TF to build. It appears CI env does not have that env var, so this is
probably fine.

mhong commented on include/swift/SIL/GraphOperationInfo.h:
```diff
@@ -188,10 +189,59 @@ struct GraphOperationInfo {
+// other TF ops (e.g. a "Const" op).
+static const char SHAPE_ARRAY_ATTR[] = "__shapes";
+
+/// Returns true if `attrName` is SHAPE_ARRAY_ATTR, `attrValue` is an array of
```
Done

Commit:
Addressed comments.

Commit:
Merge branch 'tensorflow' of github.com:apple/swift into irgen_8_attrs_input_list

mhong commented:
@swift-ci please clean test tensorflow

rxwei reviewed (approved):
LGTM

mhong commented:
Thanks for the review!

With the new TF repo bits, Linux GPU no longer builds: https://ci-external.swift.org/view/Pull%20Request/job/swift-PR-TensorFlow-Linux-gpu/155/console. I'm checking with @m3bm3b on the env vars: perhaps we'll need to turn off "export TF_NCCL_VERSION=1"

mhong commented:
@swift-ci Please test tensorflow Linux GPU

mhong commented:
@swift-ci Please clean test tensorflow Linux GPU

mhong commented:
After removing the above env var, the linux GPU test now passed: https://ci-external.swift.org/view/Pull%20Request/job/swift-PR-TensorFlow-Linux-gpu/157/
